pipeline/frontend/yaml/compiler/settings/params.go

```diff
@@ -0,0 +1,219 @@
+// Copyright 2022 Woodpecker Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package settings
+
+import (
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"codeberg.org/6543/go-yaml2json"
+	"gopkg.in/yaml.v3"
+)
+
+// ParamsToEnv uses reflection to convert a map[string]interface to a list
+// of environment variables.
+func ParamsToEnv(from map[string]interface{}, to, secrets map[string]string) (err error) {
+	if to == nil {
+		return fmt.Errorf("no map to write to")
+	}
+	for k, v := range from {
+		if v == nil || len(k) == 0 {
+			continue
+		}
+		to[sanitizeParamKey(k)], err = sanitizeParamValue(v, secrets)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// format the environment variable key
+func sanitizeParamKey(k string) string {
+	return "PLUGIN_" + strings.ToUpper(
+		strings.ReplaceAll(strings.ReplaceAll(k, ".", "_"), "-", "_"))
+}
+
+// indicate if a data type can be turned into string without encoding as json
+func isComplex(t reflect.Kind) bool {
+	switch t {
+	case reflect.Bool,
+		reflect.String,
+		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Float32, reflect.Float64:
+		return false
+	default:
+		return true
+	}
+}
+
+// sanitizeParamValue return the value of a setting as string prepared to be injected as environment variable
+func sanitizeParamValue(v interface{}, secrets map[string]string) (string, error) {
+	t := reflect.TypeOf(v)
+	vv := reflect.ValueOf(v)
+
+	switch t.Kind() {
+	case reflect.Bool:
+		return strconv.FormatBool(vv.Bool()), nil
+
+	case reflect.String:
+		return vv.String(), nil
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return fmt.Sprintf("%v", vv.Int()), nil
+
+	case reflect.Float32, reflect.Float64:
+		return fmt.Sprintf("%v", vv.Float()), nil
+
+	case reflect.Map:
+		switch v := v.(type) {
+		// gopkg.in/yaml.v3 only emit this map interface
+		case map[string]interface{}:
+			// check if it's a secret and return value if it's the case
+			value, isSecret, err := injectSecret(v, secrets)
+			if err != nil {
+				return "", err
+			} else if isSecret {
+				return value, nil
+			}
+		default:
+			return "", fmt.Errorf("could not handle: %#v", v)
+		}
+
+		return handleComplex(vv.Interface(), secrets)
```

Comment:
So this will be reached in case `isSecret=false` is returned by `injectSecret(v, secrets)`?

Reply:
yes